### PR TITLE
[build] Add `make jenkins` target.

### PR DESCRIPTION
**Pull request by jonpryor**
The `make jenkins` [^0] target is for use by Continuous Integration
machines, to build *everything* [^1].

This is expected to take an eternity. Think *hours*.

	$ time make jenkins
	...
	real	130m11.608s
	user	97m22.220s
	sys	18m20.522s

Of particular note is that the above "everything" includes
*Release configuration builds* of everything, which is something that
didn't actually work before. (Oops.)

Bump Java.Interop so that it supports building the Release
configuration, update Xamarin.Android.sln so that all required
projects are part of the Release configuration, and update
Xamarin.Android.Build.Tasks.csproj so that `JCW_ONLY_TYPE_NAMES` isn't
defined, as this was preventing compilation.

Fix **strip**(1) use: `mono-runtimes.targets` was trying to use
`strip -S` on macOS, but the value of `%(_MonoRuntime.Strip)` was
quoted, and thus attempted to execute `"strip -S" ...`, which failed.
Move the `-S` into a new `%(_MonoRuntime.StripFlags)` value.

Fixup `mono-runtimes.targets` and related files so that
`$(MonoSourceFullPath)` is used instead of a relative path. This helps
alleviate the "mental math" of determining the relative path to the
Mono checkout. Plus, the Mono checkout is supposed to be overridable,
e.g. commit d205cab2, and using `$(MonoSourceFullPath)` supports that.

Download and install `android.jar` for all supported API levels.

Fix the `Mono.Android.csproj` build so that `Mono.Android.dll` is
stored in a per-API-level intermediate directory. Otherwise, if e.g.
API-10 is built after API-23, the API-23 version will be installed,
but the JCW build will fail. Additionally, API-24 requires using
`javac -source 1.8 -target 1.8`, not 1.6.

Fix `Mono.Android/metadata` to use the correct `merge.SourceFile`
filename of `Profiles/api-24.xml.in`. Without that fix, API-24 won't
build because `NumericShaper.GetContextualShaper()` is emitted twice,
and the C# compiler doesn't like that.

Disable use of `-lz` when building for Windows. Windows doesn't
contain a `Z.DLL` to link against.

[^0]: https://en.wikipedia.org/wiki/Leeroy_Jenkins
[^1]: https://www.youtube.com/watch?v=hooKVstzbz0

### CODE REVIEW

File: build-tools/scripts/BuildEverything.mk [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# $(ALL_API_LEVELS) and $(ALL_FRAMEWORKS) must be kept in sync w/ each other
+ALL_API_LEVELS    = 1 2 3 4 5 6 7 8 9 10    11  12  13  14  15      16    17    18    19    20        21    22    23    24
```
Comment (grendello) [inline build-tools/scripts/BuildEverything.mk]: Do we really need anything below 9 here?

Comment (jonpryor) [inline build-tools/scripts/BuildEverything.mk]: Yes, for the same reason that `$(ALL_FRAMEWORKS)` has a bunch of `-` values.

That reason being that `$(FRAMEWORKS)` and others use the [**$(word n,text)** function](https://www.gnu.org/software/make/manual/html_node/Text-Functions.html#index-word), and **$(word)** requires 1-based values. If `$(ALL_API_LEVELS)` *doesn't* have 1..9, then `$(word 10, $(ALL_API_LEVELS)` won't be 10, which means `$(word 10, $(ALL_FRAMEWORKS)` won't be `v2.3`, which greatly hampers the ability, within make, to associate those values.

**Comment by dellis1972** [comment]
@jonpryor @grendello looks ok to me.

**Comment by grendello** [comment]
It looks good to me, except for one thing - unless I'm mistaken, the change to `android-toolchain.projitems` will cause a simple `make` to download and build all the platforms. Can we have a filter on API level to narrow this set to, by default, just the latest platform?

**Comment by jonpryor** [comment]
@grendello: That is correct.

Should that be fixed before merging this PR? Or done separately? (I'd *really* like to get this in so that @duncanmak can start using it...)

**Comment by grendello** [comment]
I think separately is fine

**Pull request merged**